processor/src/trace/virtual_bus/sum_check/prover/mod.rs

```diff
@@ -0,0 +1,289 @@
+        let max_degree = composition_poly.max_degree();
+        let eval_domain = EvaluationDomain::new(max_degree);
+
+        Self {
+            composition_poly,
+            eval_domain,
+            _challenger: PhantomData,
+        }
+    }
+
+    /// Given an initial claim `claim`, a mutable vector of multi-linear polynomials `mls` and
+    /// a number of rounds `num_rounds`, computes `num_rounds` iterations of the sum-check protocol
+    /// starting from claim `claim`.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - No multi-linears were provided.
+    /// - Number of rounds is zero or is greater than the number of variables of the multilinears.
+    /// - The provided multi-linears have different arities.
+    pub fn prove(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinear<E>],
+        num_rounds: usize,
+        coin: &mut C,
+        //) -> Result<(RoundClaim<E>, Proof<E>), Error> {
+    ) -> Result<Proof<E>, Error> {
+        // there should be at least one multi-linear polynomial provided
+        if mls.is_empty() {
+            return Err(Error::NoMlsProvided);
+        }
+
+        // there should be at least one round to prove
+        if num_rounds == 0 {
+            return Err(Error::NumRoundsZero);
+        }
+
+        // there can not be more rounds than variables of the multi-linears
+        let ml_variables = mls[0].num_variables();
+        if num_rounds > ml_variables {
+            return Err(Error::TooManyRounds);
+        }
+
+        // there should at least be one variable for the protocol to be non-trivial
+        if ml_variables < 2 {
+            return Err(Error::AtLeastOneVariable);
+        }
+
+        // all multi-linears should have the same arity
+        if !mls.iter().all(|ml| ml.num_variables() == ml_variables) {
+            return Err(Error::MlesDifferentArities);
+        }
+
+        let mut round_proofs = vec![];
+
+        // setup first round claim
+        let mut current_round_claim = RoundClaim {
+            eval_point: vec![],
+            claim,
+        };
+
+        // run the first round of the protocol
+        let mut round_proof = sumcheck_round(&self.composition_poly, mls);
+        round_proofs.push(round_proof);
+        // reseed with the s_0 polynomial
+        coin.reseed(H::hash_elements(&round_proofs[0].poly_evals));
+
+        for i in 1..num_rounds {
+            // generate random challenge r_i for the i-th round
+            let round_challenge = coin.draw().map_err(|_| Error::FailedToGenerateChallenge)?;
+
+            // compute the new reduced round claim
+            let new_round_claim = reduce_claim(
+                &self.eval_domain,
+                &round_proofs[i - 1],
+                current_round_claim,
+                round_challenge,
+            );
+
+            // fold each multi-linear using the round challenge
+            mls.iter_mut().for_each(|ml| ml.bind_assign(round_challenge));
```

Comment:
Question: are there benefits to binding on each iteration, as opposed to keeping the initial multilinears as-is, and prepending the previous rounds challenges `[r0, ..., rm]` to the input? I'm not sure of the relative costs of binding vs running `sumcheck_round` on more variables.

Reply:
I am not sure about the meaning of prepending the previous round challenges

Reply:
Basically, we'd leave the multilinears untouched, and call them e.g. as `ml.evaluate([r0, ..., rm, x])`, where `r0, ... rm` are the previous rounds' claims.

Basically, run the algorithm as described in the literature - keeping the `r_i`s around, and no binding.

Reply:
Ok, and what are we comparing here exactly?

> I'm not sure of the relative costs of binding vs running sumcheck_round on more variables.